cephadm: make rgw perf cache and it's size configurable

avanthakkar opened the pull request:
With the introduction of rgw perf counters cache with this PR https://github.com/ceph/ceph/pull/53003/ , add them to RGWSpec to make them configurable for rgw service deployment.


Fixes: https://tracker.ceph.com/issues/62877
Signed-off-by: avanthakkar <avanjohn@gmail.com>




<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

avanthakkar commented:
After #53003

alimaredia commented:
@avanthakkar this commit (https://github.com/alimaredia/ceph/commit/6d2f23b0f08f2e443e8594275982c5d567c9390e) should be able to configure ceph-adm  based on the rgw perf counter cache specific config variable changes made to https://github.com/ceph/ceph/pull/53003.

I'll let you take a look and apply or tweak accordingly

avanthakkar commented:
> @avanthakkar this commit ([alimaredia@6d2f23b](https://github.com/alimaredia/ceph/commit/6d2f23b0f08f2e443e8594275982c5d567c9390e)) should be able to configure ceph-adm based on the rgw perf counter cache specific config variable changes made to #53003.
> 
> I'll let you take a look and apply or tweak accordingly

Thanks! Added the changes to a new commit.

alimaredia commented on src/pybind/mgr/cephadm/services/cephadmservice.py:
```diff
@@ -1030,6 +1035,52 @@ def prepare_create(self, daemon_spec: CephadmDaemonDeploySpec) -> CephadmDaemonD
         })
 
+        mgr_map = self.mgr.get('mgr_map')
+        if 'prometheus' in mgr_map.get('services', {}):
```
I discussed this PR with @cbodley. Could we remove this assumption that the caches are always configured if the `prometheus` service is in up? We don't want the caches ever being enabled without users explicitly configuring them.

avanthakkar replied:
Ok cool! I made it general for any monitoring user but yes this also makes sense. We should somehow advice user to create rgw service creation with or w/o user/bucket cache as per their use case, probably somewhere in doc cephadm docs? wdyt @alimaredia @adk3798

cbodley commented on src/pybind/mgr/cephadm/services/cephadmservice.py:
```diff
@@ -1030,6 +1035,37 @@ def prepare_create(self, daemon_spec: CephadmDaemonDeploySpec) -> CephadmDaemonD
+            ret, out, err = self.mgr.check_mon_command({
+                'prefix': 'config set',
+                'who': utils.name_to_config_section(daemon_spec.name()),
+                'name': 'rgw_user_counters_cache',
+                'value': spec.rgw_user_counters_cache_size,
+            })
+        if spec.rgw_bucket_counters_cache:
+            ret, out, err = self.mgr.check_mon_command({
+                'prefix': 'config set',
+                'who': utils.name_to_config_section(daemon_spec.name()),
+                'name': 'rgw_bucket_counters_cache',
+                'value': spec.rgw_bucket_counters_cache_size,
```
typos? the `value` fields have `_size` at the end

avanthakkar replied:
Yes, fixed it. Thanks for noticing!

avanthakkar commented:
jenkins test api

adk3798 commented:
https://pulpito.ceph.com/adking-2023-10-01_19:14:56-orch:cephadm-wip-adk-testing-2023-10-01-1310-distro-default-smithi/

reruns of failed/dead jobs: https://pulpito.ceph.com/adking-2023-10-01_22:13:42-orch:cephadm-wip-adk-testing-2023-10-01-1310-distro-default-smithi/

After reruns, 2 failures

- https://tracker.ceph.com/issues/59704, known issue
- https://tracker.ceph.com/issues/62958, known issue


So, this PR doesn't break anything from an orch POV (at least since none of our tests actually use the new fields). Looks like the PR this depends on still hasn't been merged though.

cbodley reviewed (commented):
is there any documentation of these spec file options that need to be updated?

avanthakkar commented:
jenkins test make check

avanthakkar commented:
jenkins test dashboard cephadm

alimaredia commented:
@avanthakkar @cloudbehl any update on getting this PR merged?

avanthakkar commented:
jenkins test make check

avanthakkar commented:
@adk3798 I've fixed the failure for make check now, can you review it?

adk3798 commented on src/pybind/mgr/cephadm/services/cephadmservice.py:
```diff
@@ -1030,6 +1030,37 @@ def prepare_create(self, daemon_spec: CephadmDaemonDeploySpec) -> CephadmDaemonD
+                'prefix': 'config set',
+                'who': utils.name_to_config_section(daemon_spec.name()),
+                'name': 'rgw_user_counters_cache',
+                'value': str(spec.rgw_user_counters_cache).lower(),
```
can this ever be anything other than "true" if we get here? This variable is an optional bool, so as strings it could be "true", "false", or "none", but of those only True would pass the `if` check. If that is the case, I think we ought to just replace `str(spec.rgw_user_counters_cache).lower()` with `"true"`. Same would go for the bucket version.

adk3798 commented on src/pybind/mgr/cephadm/services/cephadmservice.py:
```diff
@@ -1030,6 +1030,37 @@ def prepare_create(self, daemon_spec: CephadmDaemonDeploySpec) -> CephadmDaemonD
+        if spec.rgw_bucket_counters_cache_size:
+            ret, out, err = self.mgr.check_mon_command({
+                'prefix': 'config set',
+                'who': utils.name_to_config_section(daemon_spec.name()),
```
we're generating `utils.name_to_config_section(daemon_spec.name())` 5 times in this `prepare_create` function (1 from before this PR, 4 being added).  Can we just have this stored in a variable before the first use case and just reference the variable 5 times instead?

adk3798 reviewed (commented):
A couple minor things. Also, I think the first 3 commits of this PR should probably be squashed together, given they're all effectively part of the same thing.

avanthakkar commented:
@adk3798  Addressed the code suggestions & seems like we have CI passing, given cephadm e2e failures are unrelated.
I see you still have testing label, are you still testing it?
